test/server.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package test supplies a fake Mixer server for use in testing. It should NOT
+// be used outside of testing contexts.
+package test // import "istio.io/mixer/test"
+
+import (
+	"fmt"
+	"net"
+
+	rpc "github.com/googleapis/googleapis/google/rpc"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	mixerpb "istio.io/api/mixer/v1"
+	"istio.io/api/mixer/v1/attributes"
+	"istio.io/mixer/pkg/attribute"
+)
+
+// AttributesServer implements the Mixer API to send mutable attributes bags to
+// a channel upon API requests. This can be used for tests that want to exercise
+// the Mixer API and validate server handling of supplied attributes.
+type AttributesServer struct {
+	GlobalDict map[string]int32
+	Attributes chan attribute.Bag
+}
+
+// NewAttributesServer creates an AttributesServer with the channel provided. If
+// a nil channel is provided, one will be created with default len.
+func NewAttributesServer(c chan attribute.Bag) *AttributesServer {
+	list := attributes.GlobalList()
+	globalDict := make(map[string]int32, len(list))
+	for i := 0; i < len(list); i++ {
+		globalDict[list[i]] = int32(i)
+	}
+	if c == nil {
+		return &AttributesServer{globalDict, make(chan attribute.Bag)}
+	}
+	return &AttributesServer{globalDict, c}
+}
+
+// Check sends a copy of the protocol buffers attributes wrapper to the server
+// channel and returns an OK response.
+func (a *AttributesServer) Check(ctx context.Context, req *mixerpb.CheckRequest) (*mixerpb.CheckResponse, error) {
+
+	requestBag := attribute.NewProtoBag(&req.Attributes, a.GlobalDict, attributes.GlobalList())
+	defer requestBag.Done()
+	a.Attributes <- attribute.CopyBag(requestBag)
+
+	resp := &mixerpb.CheckResponse{
+		Precondition: mixerpb.CheckResponse_PreconditionResult{
+			Status:        rpc.Status{Code: int32(rpc.OK)},
+			ValidUseCount: 1,
+			Attributes:    req.Attributes,
+		},
+	}
+	return resp, nil
```

Comment:
Istio/proxy integration test expects  test_mix to perform followings
1) all attributes have been processed into attributeBug.
2) for Report batch,   delta update has been processed, multiple Report should be called for its own attributeBug
3) Quota is called separately from Check.

We basically intercept at adapterManager.AspectDispatcher interface.

https://github.com/istio/mixer/blob/master/pkg/api/grpcServer.go#L45

Here is istio/ proxy intergration code.

https://github.com/istio/proxy/blob/master/src/envoy/mixer/integration_test/mixer_server.go